### PR TITLE
Save level stats separately for each level

### LINKED ISSUE #336: Feature request: save level stats to be used later
Right now we can edit the gameflow file to determine when the game should show stats screen. However, if you command it to show for example level_1 stats after completing level 3, they're incorrect – probably because they're not saved anywhere and are overwritten by the new level.

This feature would be useful mainly for custom levels, but it can be used in main game too. For example, you could make it so that after completing the final level, the game would show statistics for every level, one after another, to short of recap the experience.

This might easily be too hard to achieve for something so "meaningless" but I'd definitely have use for it...

### PR DESCRIPTION
**Pull request by rr-**
#### Checklist

- [X] I have read the [coding conventions](https://github.com/rr-/Tomb1Main/blob/master/CONTRIBUTING.md#coding-conventions)
- [X] I have added a changelog entry about what my pull request accomplishes

#### Descripton

Resolves #336.

It changes the BSON format in an incompatible way, so BSON saves made on the current master will cease to work.

#### Testing

Make the following change to the gameflow:

```diff
diff --git a/bin/cfg/Tomb1Main_gameflow.json5 b/bin/cfg/Tomb1Main_gameflow.json5
index 541304a3..e1458f7d 100644
--- a/bin/cfg/Tomb1Main_gameflow.json5
+++ b/bin/cfg/Tomb1Main_gameflow.json5
@@ -104,6 +104,7 @@
                 {"type": "start_game"},
                 {"type": "loop_game"},
                 {"type": "stop_game"},
+                {"type": "level_stats", "level_id": 1},
                 {"type": "level_stats", "level_id": 2},
                 {"type": "exit_to_level", "level_id": 3},
             ],
```

this will display stats for both level 1 and level 2 at the end of level 2.

- [x] Start the game. Pick up an item. Finish the level. It should show 1 out of 7 pickups.
- [x] Start the game. Pick up an item. Finish the level. Kill 2 wolves. Finish the level. It should show 1 out of 7 pickups for Caves and then 2 out of 29 kills for Vilca.
- [ ] Start the game. Pick up an item. Save the game using the new format. Load the game. Finish the level. It should show 1 out of 7 pickups.
- [ ] Start the game. Pick up an item. Save the game using the new format. Load the game. Finish the level. Kill 2 wolves. Finish the level. It should show 1 out of 7 pickups for Caves and then 2 out of 29 kills for Vilca.
- [ ] Start the game. Pick up an item. Finish the level. Save the game using the new format. Load the game. Kill 2 wolves. Finish the level. It should show 1 out of 7 pickups for Caves and then 2 out of 29 kills for Vilca.
- [ ] Start the game. Pick up an item. Finish the level. Kill 2 wolves. Save the game using the new format. Load the game. Finish the level. It should show 1 out of 7 pickups for Caves and then 2 out of 29 kills for Vilca.
- [x] Start the game. Pick up an item. Save the game using the legacy format. Load the game. Finish the level. It should show 1 out of 7 pickups.
- [x] Start the game. Pick up an item. Save the game using the legacy format. Load the game. Finish the level. Kill 2 wolves. Finish the level. It should show whatever for Caves and then 2 out of 29 kills for Vilca.
- [x] Start the game. Pick up an item. Finish the level. Save the game using the legacy format. Load the game. Kill 2 wolves. Finish the level. It should show whatever for Caves and then 2 out of 29 kills for Vilca.
- [x] Start the game. Pick up an item. Finish the level. Kill 2 wolves. Save the game using the legacy format. Load the game. Finish the level. It should show whatever for Caves and then 2 out of 29 kills for Vilca.

TLDR: historic levels should always work in the BSON format, they may not work for legacy saves. Current level should *always* work. Pay attention to the max counts.

In order to save and load legacy saves we need to change `allow_load` and `allow_save` members in `savegame.c`.

### CODE REVIEW

File: src/game/savegame.c [hunk 1]
```diff
@@ -167,30 +167,27 @@ static void SaveGame_LoadPostProcess()
     g_Lara.LOT.target_box = NO_BOX;
 }
 
-void InitialiseStartInfo()
+void SaveGame_InitStartEndInfo()
```
Comment (walkawayy) [inline src/game/savegame.c:170]: You changed this function name but `InitialiseStartInfo` is still called in `inventry.c` twice and `shell.c` once and it not longer exists.

**Review by walkawayy — CHANGES_REQUESTED** [review]
There are build errors in its current form.

File: src/game/game.c [hunk 2]
```diff
@@ -48,6 +48,17 @@ bool StartGame(int32_t level_num, GAMEFLOW_LEVEL_TYPE level_type)
 
 int32_t StopGame()
 {
+    SaveGame_PersistGameToEndInfo(g_CurrentLevel);
```
[the remark this answers is not in the document]
Comment (rr-) [reply src/game/game.c:51]: It's declared in savegame_common.h which is included in savegame.h.

[the remark this answers is not in the document]
Comment (walkawayy) [reply src/game/game.c:51]: I only see `void SaveGame_PersistGameToStartInfo(int level_num);` in that .h. Not `ToEndInfo`.

File: src/game/savegame.c (hunk 1 above)
Comment (rr-) [reply src/game/savegame.c:170]: Fixed

**Comment by walkawayy** [comment]
Would you mind publishing a build zip here for @Richard-L to test?

**Comment by walkawayy** [comment]
Looks like it's working to me.

image

image

image

image

**Comment by Richard-L** [comment]
Seems to be working fine for me too.

[Tomb1Main_gameflow.zip](https://github.com/rr-/Tomb1Main/files/8028210/Tomb1Main_gameflow.zip)

https://user-images.githubusercontent.com/4934209/153092963-286dc9be-c469-4616-901f-e7d8e7b58153.mp4

**Comment by rr-** [comment]
Has every scenario been tested? Esp. around legacy saves that they don't crash or do some bonkers things.

**Comment by walkawayy** [comment]
Rapora tested:
If you try to make the game show stats for a level that isn't played yet (like for level 3 at the end of level 2), it says "0 of 0" to everything. Probably not something that needs to or even can easily be fixed? It might have some very niche use in a custom level but that's it.

But I don't think that's a bug if 0 stats are shown for a level that hasn't happened yet.

**Comment by rr-** [comment]
Yeah I'd say this is such an extreme corner case that we don't need to bother.

**Comment by walkawayy** [comment]
I built and tested the legacy scenarios. It seems to work perfectly fine for current levels. Previous levels seems to have stored stats for any level in the current playthrough and if they were completed in previous game session, all the stats are 0 which IMO is correct behavior.

**Comment by rr-** [comment]
Merged manually.
